packages/backend-core/src/cache/docWritethrough.ts

```diff
@@ -0,0 +1,140 @@
+async function getCache() {
+  if (!CACHE) {
+    const client = await getDocWritethroughClient()
+    CACHE = new BaseCache(client)
+  }
+  return CACHE
+}
+
+interface ProcessDocMessage {
+  tenantId: string
+  dbName: string
+  docId: string
+  cacheKeyPrefix: string
+}
+
+export const docWritethroughProcessorQueue = createQueue<ProcessDocMessage>(
+  JobQueue.DOC_WRITETHROUGH_QUEUE
+)
+
+docWritethroughProcessorQueue.process(async message => {
+  const { tenantId, cacheKeyPrefix } = message.data
+  await context.doInTenant(tenantId, async () => {
+    const lockResponse = await locks.doWithLock(
+      {
+        type: LockType.TRY_ONCE,
+        name: LockName.PERSIST_WRITETHROUGH,
+        resource: cacheKeyPrefix,
+        ttl: 15000,
+      },
+      async () => {
+        await persistToDb(message.data)
+        console.log("DocWritethrough persisted", { data: message.data })
+      }
+    )
+
+    if (!lockResponse.executed) {
+      if (
+        lockResponse.reason !==
+        locks.UnsuccessfulRedlockExecutionReason.LockTakenWithTryOnce
+      ) {
+        console.error("Error persisting docWritethrough", {
+          data: message.data,
+        })
+        throw "Error persisting docWritethrough"
+      }
+      console.log(`Ignoring redlock conflict in write-through cache`)
+    }
+  })
+})
+
+export async function persistToDb({
+  dbName,
+  docId,
+  cacheKeyPrefix,
+}: {
+  dbName: string
+  docId: string
+  cacheKeyPrefix: string
+}) {
+  const cache = await getCache()
+
+  const db = dbUtils.getDB(dbName)
+  let doc: AnyDocument | undefined
+  try {
+    doc = await db.get(docId)
+  } catch {
+    doc = { _id: docId }
+  }
+
+  const keysToPersist = await cache.keys(`${cacheKeyPrefix}:data:*`)
+  for (const key of keysToPersist) {
+    const data = await cache.get(key, { useTenancy: false })
+    doc[data.key] = data.value
+  }
+
+  await db.put(doc)
+
+  for (const key of keysToPersist) {
+    await cache.delete(key, { useTenancy: false })
+  }
+}
+
+export class DocWritethrough {
+  private db: Database
+  private _docId: string
+  private writeRateMs: number
+  private tenantId: string
+
+  private cacheKeyPrefix: string
+
+  constructor(db: Database, docId: string, writeRateMs: number) {
+    this.db = db
+    this._docId = docId
+    this.writeRateMs = writeRateMs
+    this.cacheKeyPrefix = `${this.db.name}:${this.docId}`
+    this.tenantId = context.getTenantId()
+  }
+
+  get docId() {
+    return this._docId
+  }
+
+  async patch(data: Record<string, any>) {
+    const cache = await getCache()
+
+    await this.storeToCache(cache, data)
+
+    docWritethroughProcessorQueue.add(
+      {
+        tenantId: this.tenantId,
+        dbName: this.db.name,
+        docId: this.docId,
+        cacheKeyPrefix: this.cacheKeyPrefix,
+      },
+      {
+        delay: this.writeRateMs,
+        jobId: this.cacheKeyPrefix,
+        removeOnFail: true,
+        removeOnComplete: true,
+      }
+    )
+  }
```

Comment:
Does reusing the same `jobId` overwrite the job in Bull? If so, won't this continually push the job further forward in time? That would make the name `writeRateMs` misleading, because that's not the rate we're writing to the DB at, it is the amount of time that must pass _without_ a write for the cache to flush to the BD. It also creates a situation where particularly hot documents might not get written to the DB for a long time.

If reusing the `jobId` doesn't overwrite the job, won't we be writing to the DB for every `patch`? Is that not the situation we're trying to avoid in the first place?

Reply:
No, job id is used as idempotency key. So no 2 jobs with the same id might exist. Without the flag this would be "forever", so the same message cannot be reprocessed, but with the `removeOnX` we clean them, allowing newer messages to be queued.
Even so, this was removed after our last discussion